sdk/clientcore/core/src/main/java/io/clientcore/core/models/CoreException.java

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package io.clientcore.core.models;
+
+import io.clientcore.core.implementation.http.RetryUtils;
+
+import java.io.UncheckedIOException;
+
+/**
+ * Base class for all exceptions thrown by the client libraries and top-level core components.
+ * <p>This class extends the {@code RuntimeException} class, which means that it is an unchecked exception.</p>
+ *
+ * Instances of this class or its subclasses are typically thrown in response to errors that occur when interacting
+ * with remote services. For example, if a network request to a remote service fails, a {@code CoreException} might
+ * be thrown.
+ * It also wraps local IO exceptions and deserialization errors.
+ * <p>
+ * Client libraries should implement their own exceptions that extend this class.
+ */
+public class CoreException extends RuntimeException {
+    /**
+     * Translates a {@link Throwable} into a {@link CoreException}.
+     *
+     * @param cause the {@link Throwable} to translate
+     * @return the {@link CoreException} that was created
+     */
+    public static CoreException from(Throwable cause) {
+        return from(null, cause);
+    }
+
+    /**
+     * Translates a {@link Throwable} into a {@link CoreException}.
+     *
+     * @param message the message to use instead of the cause's message
+     * @param cause the {@link Throwable} to translate
+     * @return the {@link CoreException} that was created
+     */
+    public static CoreException from(String message, Throwable cause) {
```

Comment:
Any reason to use static factory methods to create the exception instead of using the conventional constructors to create exceptions?

Reply:
not a strong opinion, I'd like to:

- avoid someone creating new instances of base `CoreException` - client libs should create their own if they throw them
- avoid creating new instances that wrap each other. 

Factory methods seems to achieve both needs. I think it might also help us with log-and-throw.